lightning/src/chain/mod.rs

```diff
@@ -9,6 +9,117 @@
+/// funds in the channel. See [`ChannelMonitorUpdateErr`] for more details about how to handle
+/// multiple instances.
+///
+/// [`ChannelMonitor`]: channelmonitor/struct.ChannelMonitor.html
+/// [`ChannelMonitorUpdateErr`]: channelmonitor/enum.ChannelMonitorUpdateErr.html
+/// [`PermanentFailure`]: channelmonitor/enum.ChannelMonitorUpdateErr.html#variant.PermanentFailure
+pub trait Watch: Send + Sync {
+	/// Keys needed by monitors for creating and signing transactions.
+	type Keys: ChannelKeys;
+
+	/// Watches a channel identified by `funding_txo` using `monitor`.
+	///
+	/// Implementations are responsible for watching the chain for the funding transaction along
+	/// with any spends of outputs returned by [`get_outputs_to_watch`]. In practice, this means
+	/// calling [`block_connected`] and [`block_disconnected`] on the monitor.
+	///
+	/// [`get_outputs_to_watch`]: channelmonitor/struct.ChannelMonitor.html#method.get_outputs_to_watch
+	/// [`block_connected`]: channelmonitor/struct.ChannelMonitor.html#method.block_connected
+	/// [`block_disconnected`]: channelmonitor/struct.ChannelMonitor.html#method.block_disconnected
+	fn watch_channel(&self, funding_txo: OutPoint, monitor: ChannelMonitor<Self::Keys>) -> Result<(), ChannelMonitorUpdateErr>;
+
+	/// Updates a channel identified by `funding_txo` by applying `update` to its monitor.
+	///
+	/// Implementations must call [`update_monitor`] with the given update. See
+	/// [`ChannelMonitorUpdateErr`] for invariants around returning an error.
+	///
+	/// [`update_monitor`]: channelmonitor/struct.ChannelMonitor.html#method.update_monitor
+	/// [`ChannelMonitorUpdateErr`]: channelmonitor/enum.ChannelMonitorUpdateErr.html
+	fn update_channel(&self, funding_txo: OutPoint, update: ChannelMonitorUpdate) -> Result<(), ChannelMonitorUpdateErr>;
+
+	/// Returns any monitor events since the last call. Subsequent calls must only return new
+	/// events.
+	fn release_pending_monitor_events(&self) -> Vec<MonitorEvent>;
+}
+
+/// The `Filter` trait defines behavior for indicating chain activity of interest pertaining to
+/// channels.
+///
+/// This is useful in order to have a [`Watch`] implementation convey to a chain source which
+/// transactions to be notified of. Notifying may take the form of pre-filtering blocks or, in the
+/// case of [BIP 157]/[BIP 158], only fetching a block if the compact filter matches. After a
+/// transaction or an output has been registered, subsequent block retrievals from the chain source
+/// must not exclude any transactions matching the new criteria.
+///
+/// Note that use as part of a [`Watch`] implementation involves reentrancy. Therefore, these
+/// functions should not block on I/O. Implementations should instead queue the newly monitored data
+/// to be processed later. This may also require returning [`TemporaryFailure`] from `Watch`
+/// functions in order to block until the data has been processed.
```

Comment:
Nice, would it be redundant to mention to call `ChannelManager::channel_monitor_updated` when data is done processing? or have a pointer to more specific instructions on this situation?